File: src-api/EduvibeBackend/src/main/java/com/example/EduvibeBackend/service/TareaService.java
package com.example.EduvibeBackend.service;

import java.util.List;

import com.example.EduvibeBackend.dto.GetTareaDTO;
import com.example.EduvibeBackend.dto.TareaDTO;
import com.example.EduvibeBackend.entities.Clase;
import com.example.EduvibeBackend.entities.User;

/**
 * Interfaz para el servicio de gestión de tareas.
 */
public interface TareaService {
    /**
     * Crea una nueva tarea.
     * 
     * @param tareaDto los datos de la tarea a crear
     * @return el DTO de la tarea creada
     */
    TareaDTO crearTarea(TareaDTO tareaDto);

    /**
     * Obtiene una tarea por su ID.
     * 
     * @param id el ID de la tarea a obtener
     * @return el DTO de la tarea obtenida
     */
    GetTareaDTO obtenerTareaPorId(Long id);

    /**
     * Edita una tarea existente.
     * 
     * @param id el ID de la tarea a editar
     * @param tareaDto los nuevos datos de la tarea
     * @return el DTO de la tarea actualizada
     */
    TareaDTO editarTarea(Long id, TareaDTO tareaDto);

    /**
     * Elimina una tarea por su ID.
     * 
     * @param id el ID de la tarea a eliminar
     */
    void eliminarTarea(Long id);

    /**
     * Obtiene todas las tareas.
     * 
     * @return una lista de DTOs de todas las tareas
     */
    List<GetTareaDTO> obtenerTodasLasTareas();

    /**
     * Crea una nueva tarea y la asigna a una clase específica.
     * 
     * @param idClase el ID de la clase a la que se asignará la tarea
     * @param tareaDto los datos de la tarea a crear
     * @return el DTO de la tarea creada
     */
    TareaDTO crearTareaAsignandoClase(Long idClase, TareaDTO tareaDto);

    /**
     * Asigna una tarea a un usuario específico.
     * 
     * @param idTarea el ID de la tarea a asignar
     * @param idUsuario el ID del usuario al que se asignará la tarea
     */
    void asignarTareaAUsuario(Long idTarea, Integer idUsuario);

    /**
     * Edita la calificación de una tarea.
     * 
     * @param idTarea el ID de la tarea a calificar
     * @param nuevaCalificacion la nueva calificación a asignar
     */
    void editarCalificacionTarea(Long idTarea, Double nuevaCalificacion);

    /**
     * Edita la solución escrita de una tarea.
     * 
     * @param id el ID de la tarea a editar
     * @param nuevaSolucionEscrita la nueva solución escrita de la tarea
     * @return el DTO de la tarea actualizada
     */
    TareaDTO editarSolucionEscrita(Long id, String nuevaSolucionEscrita);

    /**
     * Obtiene las tareas de una clase específica para un usuario específico.
     * 
     * @param clase la clase de la cual se obtendrán las tareas
     * @param user el usuario para el cual se obtendrán las tareas
     * @return una lista de DTOs de las tareas
     */
    List<GetTareaDTO> getTareasByClaseAndUser(Clase clase, User user);

    /**
     * Obtiene las tareas de una clase específica.
     * 
     * @param clase la clase de la cual se obtendrán las tareas
     * @return una lista de DTOs de las tareas
     */
    List<GetTareaDTO> getTareasByClase(Clase clase);

    /**
     * Calcula la media de las calificaciones de una clase específica.
     * 
     * @param idClase el ID de la clase para la cual se calculará la media de las calificaciones
     * @return la media de las calificaciones
     */
    Double calcularMediaCalificaciones(Long idClase);
}

File: src-frontend/EduVibeFront/src/app/services/tarea.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { TareaDto } from '../interfaces/tarea';

@Injectable({
  providedIn: 'root'
})
export class TareaService {
  private baseUrl = 'http://localhost:9090/tareas'; // URL base de tu backend

  constructor(private http: HttpClient) { }

  // Crear una nueva tarea
  crearTarea(tareaDto: TareaDto): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/crear`, tareaDto);
  }

    // Crear una nueva tarea asignando una clase
    crearTareaAsignandoClase(idClase: number, tareaDto: TareaDto): Observable<number> {
      return this.http.post<number>(`${this.baseUrl}/crear/${idClase}`, tareaDto);
    }

  // Obtener una tarea por su ID
  obtenerTareaPorId(id: number): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/${id}`);
  }

  // Editar una tarea existente
  editarTarea(id: number, tareaDto: TareaDto): Observable<any> {
    return this.http.put<any>(`${this.baseUrl}/editar/${id}`, tareaDto);
  }

  // Eliminar una tarea por su ID
  eliminarTarea(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/eliminar/${id}`);
  }

  // Obtener todas las tareas
  obtenerTodasLasTareas(): Observable<TareaDto[]> {
    return this.http.get<any[]>(`${this.baseUrl}/todos`);
  }

  // Obtener todas las tareas por ID de clase
  obtenerTareasPorClase(idClase: number): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/clase/${idClase}`);
  }
  subirArchivo(idTarea: number, archivo: File): Observable<any> {
    const formData: FormData = new FormData();
    formData.append('archivo', archivo, archivo.name);
    return this.http.post<any>(`${this.baseUrl}/${idTarea}/archivo`, formData);
  }

  descargarArchivo(idTarea: number, indice: number): Observable<Blob> {
    return this.http.get(`${this.baseUrl}/${idTarea}/archivo/${indice}`, {
      responseType: 'blob'
    });
  }

  obtenerMediaCalificaciones(idClase: number): Observable<number> {
    return this.http.get<number>(`${this.baseUrl}/mediaCalificaciones/${idClase}`);
  }

    // Editar la calificación de una tarea
    editarCalificacion(id: number, nuevaCalificacion: number): Observable<void> {
      return this.http.put<void>(`${this.baseUrl}/calificacion/${id}?nuevaCalificacion=${nuevaCalificacion}`, null);
    }

    // Obtener todas las tareas por ID de clase para el usuario actual
    obtenerTareasPorClaseParaUsuarioActual(idClase: number): Observable<any> {
      return this.http.get<any>(`${this.baseUrl}/clase/user/${idClase}`);
    }

    asignarTareaAUsuario(idTarea: number, idUsuario: number): Observable<void> {
      const url = `${this.baseUrl}/asignar?idTarea=${idTarea}&idUsuario=${idUsuario}`;
      return this.http.post<void>(url, null);
    }
}
